Added ROS 2 compatibility compilation instructions and example (#48)
eProsima/Shapes-Demo-Docs: docs/examples/ros2_compatibility.rst

.. _examples_ros2_compatibility:

ROS 2 Compatibility
===================

As stated in the :ref:`linux_ros2_compilation` section, eProsima Shapes Demo can be built and used on a ROS 2 installation with additional ROS 2 features.

> 

Reviewer: ```suggestion
```

With ROS 2 features enabled, an additional "Use ROS2 Topics" checkbox in the Participant configuration dialog will be shown.

.. image:: /01-figures/participant_ros_enabled.png
   :scale: 100 %
   :alt: ROS 2 Participant Configuration Options
   :align: center

Enabling this box will put Shapes Demo in ROS 2 mode.
ROS 2 Shapes Demo main window changes to reflect this by showing a different title and the `Vulcanexus <https://www.vulcanexus.org/>`_ logo.

.. image:: /01-figures/ros2_shapes_demo.png
   :scale: 100 %
   :alt: ROS 2 Shapes Demo Main Window
   :align: center

When using eProsima Shapes Demo in this mode, ROS 2 will be aware of the Topics transmitted.
For instance, this is the output of the ``ros2 topic list`` command.

   .. code-block:: bash

      $ ros2 topic list -t
      /Square [shapes_demo_typesupport/idl/KeylessShapeType]
      /parameter_events [rcl_interfaces/msg/ParameterEvent]

>       /rosout [rcl_interfaces/msg/Log]

Reviewer: This is not working for me. I'd split this in two to have an easy-to-copy command and the output on different boxes

Author: I believe this is a side-effect of the SHM issue between Fast DDS versions. I've reproduced the issue with the v2.9.1 (as it is on the repos file) and it's working properly when using `master` for both Shapes Demo and Fast DDS.

Reviewer: You're correct! I disabled SHM and datasharing and it works

> Since there's a TypeSupport available for these messages, it can be used by ROS 2 to interact with the different ShapesDemo topics.

Reviewer: ```suggestion
Since there is a TypeSupport available for these messages, it can be used by ROS 2 to interact with the different ShapesDemo topics.
```

For instance, assuming the Shapes Demo TypeSupport was built along with Shapes Demo and is currently available in the current installation folder, a subscription to a Topic could be made like so:

   .. code-block:: bash

>       $ source install/setup.bash

Reviewer: ```suggestion
      $ source ~/shapes_demo_ws/install/setup.bash
```

      $ ros2 topic echo /Square
      color: RED
      x: 175
      y: 215
      shapesize: 30
      ---
      color: RED
      x: 169
      y: 210
      shapesize: 30

>       ---

Reviewer: I'd split this too

Analogously, a publication could be made like this:

   .. code-block:: bash

>       $ source install/setup.bash

Reviewer: ```suggestion
      $ source ~/shapes_demo_ws/install/setup.bash
```

      $ ros2 topic pub /Square shapes_demo_typesupport/idl/KeylessShapeType "{ color: BLUE, x: 155, y: 150, shapesize: 30}"
      publisher: beginning loop

>       publishing #1: shapes_demo_typesupport.idl.KeylessShapeType(color='BLUE', x=155, y=150, shapesize=30)

Reviewer: Please split

.. _examples_ros2_compatibility:

ROS 2 Compatibility
===================

As stated in the :ref:`linux_ros2_compilation` section,
eProsima Shapes Demo can be built and used on a ROS 2 installation with additional ROS 2 features.
With ROS 2 features enabled, an additional "Use ROS2 Topics" checkbox in the Participant configuration dialog will be shown.

.. image:: /01-figures/participant_ros_enabled.png
   :scale: 100 %
   :alt: ROS 2 Participant Configuration Options
   :align: center

Enabling this box will put Shapes Demo in ROS 2 mode.
ROS 2 Shapes Demo main window changes to reflect this by showing a different title and the `Vulcanexus <https://www.vulcanexus.org/>`_ logo.

.. image:: /01-figures/ros2_shapes_demo.png
   :scale: 100 %
   :alt: ROS 2 Shapes Demo Main Window
   :align: center

When using eProsima Shapes Demo in this mode, ROS 2 will be aware of the Topics transmitted.
All topics currently known by ROS can be listed as follows:

   .. code-block:: bash

>       ros2 topic list -t

Reviewer: dedent

A typical output of the previous command with a Square created by Shapes Demo would be:

   .. code-block:: bash

> 
>       ros2 topic list -t
> 

Reviewer: ```suggestion
```

      /Square [shapes_demo_typesupport/idl/KeylessShapeType]
      /parameter_events [rcl_interfaces/msg/ParameterEvent]

>       /rosout [rcl_interfaces/msg/Log]

Reviewer: dedent

Since there is a TypeSupport available for these messages, it can be used by ROS 2 to interact with the different ShapesDemo topics.
For instance, assuming the Shapes Demo TypeSupport was built along with Shapes Demo and is currently available in the current installation folder, a subscription to a Topic could be made like so:

   .. code-block:: bash

      source ~/shapes_demo_ws/install/setup.bash
      ros2 topic echo /Square

The terminal will start showing the data transmitted on the Square topic:

   .. code-block:: bash

      color: RED
      x: 175
      y: 215
      shapesize: 30
      ---
      color: RED
      x: 169
      y: 210
      shapesize: 30
      ---

Analogously, a publication could be made like this:

   .. code-block:: bash

      source ~/shapes_demo_ws/install/setup.bash
      ros2 topic pub /Square shapes_demo_typesupport/idl/KeylessShapeType "{ color: BLUE, x: 155, y: 150, shapesize: 30}"

On successful execution, this is what would be shown on the terminal:

   .. code-block:: bash

      publisher: beginning loop

>       publishing #1: shapes_demo_typesupport.idl.KeylessShapeType(color='BLUE', x=155, y=150, shapesize=30)

Reviewer: Adding a screenshot of the blue square here would be amazing!
